lib/ansible/galaxy/api.py

```diff
@@ -46,6 +47,27 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def rate_limit_backoff(function):
```

Comment:
Any particular reason we can't use/vendor https://github.com/litl/backoff? That lib has a fantastic API and we'd not have to reimplement its slick UX.

Also, we've encountered other places where the backoff is used and is sometimes poorly implemented. I recall some modules needed it in the past. So it'd make sense to have it made available in a generic fashion as opposed to another module-specific implementation.

Reply:
I guess we could vendor it? It's such a minimal change that I didn't think it was necessary. I thought we may want to backport, and if so, we shouldn't add a new dep (that's also why things aren't user-configurable yet as that's feature-ish).

Reply:
Generally speaking, I'd like us to be very cautious on adding any new deps, and I'm not a huge fan of vendoring.  That lib also isn't available in a number of distros that I've just checked.

Reply:
we already have these type functions in module_utils/api.py as decorators

Reply:
@bcoca we don't have one with a specifiable backoff strategy or one that can handle specific error codes. Should I move this to common code? I debated over where to put it.

Reply:
yes, this is useful for other things, so i would put it under module_utils, even expand existing functions if it makes sense

Reply:
@sivel FWIW if we were to use backoff, we'd have to vendor it to make this available to modules on target hosts especially because it's useful for modules talking to services over API.

@s-hertel yep, specifically because this is useful to modules it should to under `module_utils`.

Reply:
So out of curiosity, I did a quick GH repo search and here's my findings:
* exponential backoff in modules:
  * https://github.com/ansible/ansible/blob/4c5ce5a1a9e79a845aff4978cfeb72a0d4ecf7d6/lib/ansible/modules/apt_repository.py#L645
  * https://github.com/ansible/ansible/blob/2bff120db6bf96dd0085c6ce1b29f3999b872db4/lib/ansible/modules/apt.py#L1189
* in action plugins:
  * https://github.com/ansible/ansible/blob/a51a6f4a259b45593c3f803737c6d5d847258a83/lib/ansible/plugins/action/reboot.py#L302
* in connection plugins:
  * https://github.com/ansible/ansible/blob/96905120698e3118d8bafaee5ebe8f83d2bbd607/lib/ansible/plugins/connection/psrp.py#L859 (delegated to Jordan's lib — `pypsrp`)
* in tests:
  * https://github.com/ansible/ansible/blob/a077bca5d5866603cdc4a2a13aef8f416860aacb/test/support/integration/plugins/modules/ec2_vpc_subnet.py#L259 (`AWSRetry` — looks like a post-migration leftover, not actually used)
* in `module_utils`:
  * https://github.com/ansible/ansible/blob/a077bca5d5866603cdc4a2a13aef8f416860aacb/test/support/integration/plugins/module_utils/cloud.py#L48
  * https://github.com/ansible/ansible/blob/a51a6f4a259b45593c3f803737c6d5d847258a83/lib/ansible/module_utils/api.py#L94 (not exponential but linear backoff)
* and there's even snippets in docs showing how to implement this in YAML playbooks: 
  *  https://github.com/ansible/ansible/blob/ccd9a992cf7b7b057cfd9a2d06c69c44b7094696/docs/docsite/rst/user_guide/complex_data_manipulation.rst#L28 (not related to our problem I guess)
* also, this very module has exponential backoff already baked-in in one place too:
  * https://github.com/ansible/ansible/blob/00bd0b893d5d21de040b53032c466707bacb3b93/lib/ansible/galaxy/api.py#L689

Reply:
The first two links (apt and apt_repository) will be simple to replace with a generic wrapper.

`reboot` has some debug messages during the retry process that would overcomplicate a common retry decorator, in my opinion.

The exponential backoff in the `ansible/lib/ansible/galaxy/api.py` is actually handling two things - retrying errors and waiting until the import is complete. Those are coupled together and will need some refactoring to use a common retry decorator.

I'm consolidating code with the `retry` function in `lib/ansible/module_utils/api.py` (not extending it, as it's doing some not-so-generic-things). I'll fix `apt` and `apt_repository` in a follow-up PR.

Everything else looks unrelated. The docs are great for things that don't implement retries. The stuff in `ansible/test/support` needs to stay as-it-is. `psrp` is a false positive since the option is for the python library, and not a case of Ansible reinventing the wheel.